sympy/stats/stochastic_process_types.py

```diff
@@ -623,13 +676,83 @@ class DiscreteMarkovChain(DiscreteTimeStochasticProcess, MarkovProcess):
     """
     index_set = S.Naturals0
 
-    def __new__(cls, sym, state_space=S.Reals, trans_probs=None):
+    def __new__(cls, sym, state_space=None, trans_probs=None):
         sym = _symbol_converter(sym)
-        state_space = _set_converter(state_space)
-        if trans_probs != None:
+
+        # Try to never have None as state_space or trans_probs.
+        # This helps a lot if we get it done at the start.
+        if (state_space is None) and (trans_probs is None):
+            _n = Dummy('n', integer=True, nonnegative=True)
+            state_space = Range(_n)
+            trans_probs = MatrixSymbol('_T', _n, _n)
+
+        elif state_space is None:
             trans_probs = _matrix_checks(trans_probs)
+            state_space = _state_converter(Range(trans_probs.shape[0]))
+
+        elif trans_probs is None:
+            state_space = _state_converter(state_space)
+            if isinstance(state_space, Range):
+                _n = (state_space.stop - state_space.start) // state_space.step
+            else:
+                _n = len(state_space)
+            trans_probs = MatrixSymbol('_T', _n, _n)
+
+        else:
+            state_space = _state_converter(state_space)
+            trans_probs = _matrix_checks(trans_probs)
+            # Range object doesn't want to give a symbolic size
+            # so we do it ourselves.
+            if isinstance(state_space, Range):
+                ss_size = (state_space.stop - state_space.start)//state_space.step
+            else:
+                ss_size = len(state_space)
+            if ss_size != trans_probs.shape[0]:
+                raise ValueError('The size of the state space and the number of '
+                                 'rows of the transition matrix must be the same.')
+
         return Basic.__new__(cls, sym, state_space, trans_probs)
 
+    @property
+    def index_of(self):
```

Comment:
```suggestion
    def _index_of(self):
```
Similar, is this useful to the end user who is just doing, `P` and `E` queries?

Reply:
This one should be. Probability queries could be performed such as `P(Eq(X[2],X.index_of['Sunny']),...)`. I think I'll add that to the docstring for that weather model.